amazon/ion/ion_py_objects.py

```diff
@@ -0,0 +1,586 @@
+from decimal import Decimal
+from collections.abc import MutableMapping
+
+from amazon.ion.core import IonType, IonEvent
+from amazon.ion.symbols import SymbolToken
+
+
+class IonPyNull_new(object):
+    __name__ = 'IonPyNull_new'
+    __qualname__ = 'IonPyNull_new'
+    ion_type = IonType.NULL
+
+    def __init__(self, ion_type=IonType.NULL, value=None, annotations=()):
+        self.ion_type = ion_type  # TODO initialized to NULL type first, what's the real type?
+        self.ion_annotations = annotations
+
+    def __bool__(self):
+        return False
+
+    def __copy__(self):
+        args, kwargs = self._to_constructor_args(self)
+        value = self.__class__(*args, **kwargs)
+        value.ion_type = self.ion_type
+        value.ion_annotations = self.ion_annotations
+        return value
+
+    @staticmethod
+    def _to_constructor_args(value):
+        return (None, value,), {}
+
+    @classmethod
+    def from_event(cls, ion_event):
+        if ion_event.value is not None:
+            args, kwargs = cls._to_constructor_args(ion_event.value)
+        else:
+            args, kwargs = (None, None, ()), {}
+        value = cls(*args, **kwargs)
+        value.ion_type = ion_event.ion_type
+        value.ion_annotations = ion_event.annotations
+        return value
+
+    @classmethod
+    def from_value(cls, ion_type, value, annotations=()):
+        if value is None:
+            value = IonPyNull_new()
+        else:
+            args, kwargs = cls._to_constructor_args(value)
+            value = cls(*args, **kwargs)
+        value.ion_type = ion_type
+        value.ion_annotations = annotations
+        return value
+
+    def to_event(self, event_type, field_name=None, in_struct=False, depth=None):
+        value = self
+        if isinstance(self, IonPyNull_new) or self.ion_type.is_container:
+            value = None
+
+        if in_struct:
```

Comment:
seems like we could avoid the logic and simplify the parameters by passing only passing `field_name` if it were in a struct. realize that _may_ be a breaking change, although I'd like to understand what behavior is expected when passing a field_name with in_struct=False

Reply:
I didn't spend too much effort on the `to_event` API since it's mainly used by the pure python implementation. But I took a look, it seems that we do not have to maintain `field_name`. It requires more investigation to see if this will cause breaking changes (at least some low-level internal methods need this parameter but the top-level dump doesn't).